src/main/java/org/spongepowered/common/mixin/core/server/management/MixinPlayerProfileCache.java

```diff
@@ -0,0 +1,225 @@
+
+        for (String name : names) {
+            result.put(name, Optional.ofNullable((GameProfile) this.getByNameNoLookup(name)));
+        }
+
+        return result.isEmpty() ? ImmutableMap.of() : ImmutableMap.copyOf(result);
+    }
+
+    @Override
+    public Optional<GameProfile> lookupByName(String name) {
+        SingleProfileLookupCallback callback = new SingleProfileLookupCallback();
+
+        this.getServer().getGameProfileRepository().findProfilesByNames(new String[]{name}, Agent.MINECRAFT, callback);
+
+        return callback.getResult();
+    }
+
+    @Override
+    public Map<String, Optional<GameProfile>> lookupByNames(Iterable<String> names) {
+        checkNotNull(names, "names");
+
+        Map<String, Optional<GameProfile>> result = Maps.newHashMap();
+
+        this.getServer().getGameProfileRepository().findProfilesByNames(Iterables.toArray(names, String.class), Agent.MINECRAFT, new MapProfileLookupCallback(result));
+
+        return result.isEmpty() ? ImmutableMap.of() : ImmutableMap.copyOf(result);
+    }
+
+    @Override
+    public Optional<GameProfile> fillProfile(GameProfile profile, boolean signed) {
+        checkNotNull(profile, "profile");
+
+        return Optional.ofNullable((GameProfile) this.getServer().getMinecraftSessionService().fillProfileProperties((com.mojang.authlib.GameProfile) profile, signed));
+    }
+
+    @Override
+    public Collection<GameProfile> getProfiles() {
+        return this.usernameToProfileEntryMap.values().stream()
+                .map(entry -> (GameProfile) entry.getGameProfile())
+                .collect(GuavaCollectors.toImmutableSet());
+    }
+
+    @Redirect(method = "getGameProfile(Lnet/minecraft/server/MinecraftServer;Ljava/lang/String;)Lcom/mojang/authlib/GameProfile;",
+            at = @At(value = "INVOKE", target = "Lcom/mojang/authlib/GameProfileRepository;findProfilesByNames([Ljava/lang/String;"
+                    + "Lcom/mojang/authlib/Agent;Lcom/mojang/authlib/ProfileLookupCallback;)V"))
+    private static void onGetGameProfile(GameProfileRepository repository, String[] names, Agent agent, ProfileLookupCallback callback) {
+        GameProfileCache cache = Sponge.getServer().getGameProfileManager().getCache();
+        if (cache instanceof PlayerProfileCache) {
+            repository.findProfilesByNames(names, agent, callback);
+        } else {
+            // The method we're redirecting into obtains the resulting GameProfile from
+            // the callback here.
+            callback.onProfileLookupSucceeded((com.mojang.authlib.GameProfile) cache.getOrLookupByName(names[0]).orElse(null));
+        }
+    }
+
+    @Nullable
+    public com.mojang.authlib.GameProfile getByNameNoLookup(String username) {
+        @Nullable IMixinPlayerProfileCacheEntry entry = this.usernameToProfileEntryMap.get(username.toLowerCase(Locale.ROOT));
+
+        if (entry != null && new Date().getTime() >= entry.getExpirationDate().getTime()) {
```

Comment:
You should use `System.currentTimeMillis()`. Eliminates 1 object creation and does the same (not that 1 object allocation is really bad, but might as well be clean)